ipq40xx: add support for EnGenius ENS620EXT

steveglennon opened the pull request:
Added basic support to get the device up and running for a sysupgrade    |
image only.
There is currently no way back to factory firmware, so this is a one-way
street to OpenWRT.
Install from factory condition is convoluted, and may brick your device:
1) Enable SSH and disable the CLI on the factory device from the web user
   interface (Management->Advanced)
2) Reboot the device
3) Override the default, limited SSH shell:
   a) Get into the ssh shell:
      ssh admin@192.168.1.1 /bin/sh --login
   b) Change the dropbear script to disable the limited shell. At the
      empty command prompt type:
        sed -i '/login_ssh/s/^/#/g’ dropbear
        /etc/init.d/dropbear restart
        exit
4) ssh in to a (now-) normal OpenWRT SSH session
5) Flash your built image
   a) scp openwrt-ipq40xx-engenius_ens620ext-squashfs-sysupgrade.bin
      admin@192.168.1.1:/tmp/
   b) ssh admin@192.168.1.1
   c) sysupgrade
      /tmp/openwrt-ipq40xx-engenius_ens620ext-squashfs-sysupgrade.bin
6) After flash completes (it may say "Upgrade failed" followed by
   "Upgrade completed") and device reboots, log in to newly flashed
   system. Note you will now need to ssh as root rather than admin.

Signed-off-by: Steve Glennon <s.glennon@cablelabs.com>

Thanks for your contribution to OpenWrt!

To help keep the codebase consistent and readable,
and to help people review your contribution,
we ask you to follow the rules you find in the wiki at this link
https://openwrt.org/submitting-patches

Please remove this message before posting the pull request.

steveglennon committed:
ipq40xx: add support for EnGenius ENS620EXT

Hardware
--------
CPU:   Qualcomm IPQ4018
RAM:   256M
FLASH: 32M SPI NOR W25Q256
ETH:   QCA8075
WiFi2: IPQ4018 2T2R 2SS b/g/n
WiFi5: IPQ4018 2T2R 2SS n/ac
LED:    - Power amber
        - LAN1(PoE) green
        - LAN2 green
        - Wi-Fi 2.4GHz green
        - Wi-Fi 5GHz green
BTN:    - WPS
UART:  115200n8 3.3V J1
       VCC(1) - GND(2) - TX(3) - RX(4)

Added basic support to get the device up and running for a sysupgrade
image only.
There is currently no way back to factory firmware, so this is a one-way
street to OpenWRT.
Install from factory condition is convoluted, and may brick your device:
1) Enable SSH and disable the CLI on the factory device from the web user
   interface (Management->Advanced)
2) Reboot the device
3) Override the default, limited SSH shell:
   a) Get into the ssh shell:
      ssh admin@192.168.1.1 /bin/sh --login
   b) Change the dropbear script to disable the limited shell. At the
      empty command prompt type:
        sed -i '/login_ssh/s/^/#/g’ dropbear
        /etc/init.d/dropbear restart
        exit
4) ssh in to a (now-) normal OpenWRT SSH session
5) Flash your built image
   a) scp openwrt-ipq40xx-engenius_ens620ext-squashfs-sysupgrade.bin
      admin@192.168.1.1:/tmp/
   b) ssh admin@192.168.1.1
   c) sysupgrade -n
      /tmp/openwrt-ipq40xx-engenius_ens620ext-squashfs-sysupgrade.bin
6) After flash completes (it may say "Upgrade failed" followed by
   "Upgrade completed") and device reboots, log in to newly flashed
   system. Note you will now need to ssh as root rather than admin.

Signed-off-by: Steve Glennon <s.glennon@cablelabs.com>

robimarko commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,313 @@
+		led-upgrade = &power;
+	};
+
+	soc {
```
Missing rng node here

steveglennon replied:
Fixed

robimarko commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,313 @@
+			status = "okay";
+		};
+
+		wifi@a000000 {
```
Please use &wifi0 and &wifi1, follow the example of other boards

robimarko commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,313 @@
+
+		wifi@a000000 {
+			status = "okay";
+			qcom,ath10k-calibration-variant = "RT-AC58U";
```
I am sure that this is completely wrong.
Take a look at previous ipq40xx PRs

steveglennon replied:
Fixed

robimarko commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,313 @@
+	pinctrl-names = "default";
+	status = "okay";
+	cs-gpios = <&tlmm 54 GPIO_ACTIVE_HIGH>,
+		   <&tlmm 59 GPIO_ACTIVE_HIGH>;
```
Wrong as you are only using one CS

steveglennon replied:
Fixed

robimarko commented on target/linux/ipq40xx/files-4.19/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,313 @@
+
+		mux_cs {
+			function = "gpio";
+			pins = "gpio54", "gpio59";
```
Same as below

steveglennon replied:
Fixed

robimarko commented on target/linux/ipq40xx/patches-4.14/998-arm-boot-add-ens620ext-dts-file.patch:
```diff
@@ -0,0 +1,17 @@
+ 	qcom-ipq4018-ap120c-ac.dtb \
+ 	qcom-ipq4018-ea6350v3.dtb \
+ 	qcom-ipq4018-eap1300.dtb \
++	qcom-ipq4018-ens620ext.dtb \
```
No need for another patch, edit the existing 997 patch

steveglennon replied:
I was not sure which was this should go - thanks for the guidance and I have removed the proposed patch and added the line into 901 patch

robimarko commented on target/linux/ipq40xx/patches-4.19/998-arm-boot-add-ens620ext-dts-file.patch:
```diff
@@ -0,0 +1,17 @@
+ 	qcom-ipq4018-ap120c-ac.dtb \
+ 	qcom-ipq4018-ea6350v3.dtb \
+ 	qcom-ipq4018-eap1300.dtb \
++	qcom-ipq4018-ens620ext.dtb \
```
Same as above

steveglennon replied:
Fixed

robimarko commented:
PR has multiple copy/paste things that will not work.
I have added comments to some of them but most likely there are plenty more.

Please rebase this on master as in DTS you still have qcom,ipq4018 compatible.
And take a look at previous PRs made for ipq40xx devices for proper examples and explanations for most stuff.

robimarko commented:
Please sort this out, you pulled and then pushed everything after you commit.
Please dont do this

steveglennon commented:
Sorry for the pull/push issue - sorry new to rebasing. With some more expert git help I fixed the issues you raised and squashed the empty commit.
(background: the patch was originally written late last year before some of the recent DTS work on IPQ4018, and was modeled on the rt-ac58u dts file at the time. I brought it current with the head but had not re-checked the dts files to see that some nodes had changed)
Sorry if I abuse your patience.

chunkeey commented on target/linux/ipq40xx/base-files/etc/board.d/01_leds:
```diff
@@ -35,6 +35,12 @@ engenius,eap1300)
+        ucidef_set_led_wlan "wlan2g" "WLAN2G" "${boardname}:green:wlan2G" "phy0tpt"
+        ucidef_set_led_wlan "wlan5g" "WLAN5G" "${boardname}:green:wlan5G" "phy1tpt"
+        ucidef_set_led_netdev "lan" "LAN" "${boardname}:green:lan" "eth0"
+        ucidef_set_led_netdev "zlan1" "ZLAN1" "${boardname}:green:zlan1" "eth1"
```
I'm curious, do you know what's up with the "ZLAN1" name? Is this something from the stock firmware or where is it coming from?

steveglennon replied:
(I've been been feeling my way into adding device support) Device has two lan ports and two LEDs. Was not sure if "lan" label had a special meaning so copied that for eth0, adding something else for eth1/lan1.

What is the appropriate/expected change here? lan0/LAN0/green:lan0 eth0 and lan1/LAN1/green/lan1 eth1?

chunkeey replied:
I'm sorry, I haven't noticed this earlier.

> What is the appropriate/expected change here?

As a general rule for most labels: follow the label/pictogram/sticker that written/printed/etched on the casing, silk-screened on the PCB or specified in the manual.

As for LED labels: They should follow the linux' kernels upstream recommendation of
"devicename:color:function" as per:
https://github.com/spotify/linux/blob/master/Documentation/leds-class.txt
(See section LED Device Naming)

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/base-files/etc/board.d/02_network:
```diff
@@ -40,6 +40,9 @@ ipq40xx_setup_interfaces()
 	compex,wpj428)
 		ucidef_set_interface_lan "eth0 eth1"
 		;;
+	engenius,ens620ext)
```
Looks like this could fit with the 8dev,jalapeno and alfa-network,ap120c-ac devices. Please add your device there.

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+#include <dt-bindings/soc/qcom,tcsr.h>
+
+/ {
+	model = "ENGENIUS ENS620EXT";
```
The EAP1300 uses EnGenius (which seem to match their website). Can you please change this?

steveglennon replied:
Fixed in next commit

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+		};
+	};
+
+	keys {
```
(This is optional)

"buttons" could also be a valid name for this node. Please take whatever fits the best.

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+			gpios = <&tlmm 3 GPIO_ACTIVE_HIGH>;
+		};
+
+		wan5G {
```
is the "l" missing here, shouldn't this be w**l**an5G?

steveglennon replied:
Thanks. Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+		#address-cells = <1>;
+		#size-cells = <0>;
+		/*
+		 * U-boot looks for "n25q128a11" node,
```
Is this comment left over copy-pasta from the RT-AC58U? 
If so, please remove it together with the bogus:
```linux,modalias = "m25p80", "mx25l1606e", "n25q128a11";```
down below.

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition0@0 {
```
Please change all "partition$x@$y" -> "partition@$y"

steveglennon replied:
Fixed this in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+			#size-cells = <1>;
+
+			partition0@0 {
+				label = "0:SBL1";
```
The "0:" partition name prefixes are Qualcomminnsm. Please remove them as they don't really serve any purpose and just makes the labels look odd.

steveglennon replied:
Fixed this in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+				reg = <0x00180000 0x00010000>;
+				read-only;
+			};
+			partition5@190000 {
```
Please remove the 0:HLOS partition completely. It's the qualcomm-speak for "high level os" and this is going to
clash with the "firmware" partition splitter (see denx,fit) we have.

steveglennon replied:
Fixed this in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+				label = "0:HLOS";
+				reg = <0x00190000 0x00500000>;
+			};
+			partition5@690000 {
```
Please also remove the rootfs partition. The mtd splitter will automatically generate one for you.

steveglennon replied:
I had to do this to get to the point where sysupgrade on the unmodified/factory device would successfully flash the sysupgrade image we create.

chunkeey replied:
That sounds very strange. I could see that you want to have the HLOS and rootfs partitions to make it easier to flash the original firmware, but not the other way around. I think keeping these partitions around will confuse the heck out of people. Because I don't know if you noticed, but the two mtdsplitter will make three dynamic partitions out of the single "firmware".
You can verify this by looking at your /proc/mtd on the device.

steveglennon replied:
I've made these changes and it boots fine - but I need to force the device back to factory (manual using spi flashing of copies of MTD partitions I kept) to make sure it can be upgraded using the factory image and sysupgrade.

chunkeey replied:
Hm, so you need the HLOS and rootfs to test the factory->openwrt install procedure. :relieved: 

I think you can do this with just the "firmware" partition as well. The trick is to merge the
saved HLOS and rootfs together. 
   ```cat factory-HLOS-image factory-rootfs > factory-image.bin```

And then:

   ```mtd -r write factory-image.bin firmware```
(the -r option is there to do a reboot when the mtd write finishes)

or what might also now work as well:

   ```sysupgrade -F factory-image.bin```

steveglennon replied:
When I have everything fixed from your (very helpful) comments I will re-verify that I can flash the device from factory before I push to update the pull request.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+				label = "rootfs";
+				reg = <0x00690000 0x00FD0000>;
+			};
+			partition5@1660000 {
```
I'm curious. Do you know what's in the failsafe partition? Is it another kernel+rootfs/initramfs or what does it do?

steveglennon replied:
Not sure. This is what smeminfo shows from uboot:
(IPQ40xx) # smeminfo
flash_type:             0x6
flash_index:            0x0
flash_chip_select:      0x0
flash_block_size:       0x10000
flash_density:          0x2000000
partition table offset  0x0
No.: Name             Attributes            Start             Size
  0: 0:SBL1           0x0000ffff              0x0          0x40000
  1: 0:MIBIB          0x002040ff          0x40000          0x20000
  2: 0:QSEE           0x0000ffff          0x60000          0x60000
  3: 0:CDT            0x0000ffff          0xc0000          0x10000
  4: 0:DDRPARAMS      0x0000ffff          0xd0000          0x10000
  5: 0:APPSBLENV      0x0000ffff          0xe0000          0x10000
  6: 0:APPSBL         0x0000ffff          0xf0000          0x90000
  7: 0:ART            0x0000ffff         0x180000          0x10000
  8: 0:HLOS           0x0000ffff         0x190000         0x500000
  9: rootfs           0x0000ffff         0x690000         0xfd0000
 10: failsafe         0x0000ffff        0x1660000         0x8f0000
 11: u-boot-env       0x0000ffff        0x1f50000          0x10000
 12: userconfig       0x0000ffff        0x1f60000          0xa0000

chunkeey replied:
Not much useful information there. Could you download the content of the failsafe partition to your PC and run "binwalk" on it? This should give us a overview of what's stored there.

steveglennon replied:
Tried a couple of different variants of binwalk.
Here's one version's output:
Binwalk v2.1.2-720f770
Craig Heffner, ReFirmLabs
https://github.com/ReFirmLabs/binwalk

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             device tree image (dtb)
228           0xE4            gzip compressed data, maximum compression, has original file name: "Image", from Unix, last modified: 2017-07-26 06:48:55
3252188       0x319FDC        device tree image (dtb)
3288960       0x322F80        device tree image (dtb)
3329280       0x32CD00        device tree image (dtb)
3369676       0x336ACC        device tree image (dtb)
3402876       0x33EC7C        device tree image (dtb)
3436220       0x346EBC        device tree image (dtb)
3469720       0x34F198        device tree image (dtb)

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+				label = "firmware";
+				reg = <0x00190000 0x14D0000>;
+			};
+	/* 0x00180000 - 0x00200000 unused */
```
if this is more leftover from the RT-AC58U, then please remove it.

steveglennon replied:
Fixed in next commit.

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+
+&wifi0 {
+	status = "okay";
+	qcom,ath10k-calibration-variant = "EnGenius-ENS620EXT";
```
You'll need to submit your device's boarddata files (see in the vendor firmware for files like boardData_0.bin under /lib/firmware/...) upstream to ath10k-devel mailing list.

You can find instructions and examples on the linux-wireless wiki:
https://wireless.wiki.kernel.org/en/users/drivers/ath10k/boardfiles

Once you are done, please pack the board-files into a ipq-wifi package for your device.
You can find instructions for how to do this in the ipq-wifi commit: fa03d441e96eeec0781ccbe757a3641c9dcec785

If you have any more questions regarding the boardfiles or ipq-wifi please ask.

steveglennon replied:
Thank you thank you thank you. Was not aware of how this worked. I built the board files with the qca swiss army knife and now 5GHz works properly!!! 

I'll submit the board files and the short term addition to ipq_wifi will be in the next commit.
Thanks for all the pointers - just what I needed.

chunkeey replied:
Thank you for posting the boardfiles to the ath10k mailing list.

http://lists.infradead.org/pipermail/ath10k/2019-March/013034.html

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+
+&wifi1 {
+	status = "okay";
+	qcom,ath10k-calibration-variant = "EnGenius-ENS620EXT";
```
Same as above.

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -158,6 +158,22 @@ define Device/engenius_eap1300
+        PAGESIZE := 2048
+        DTB_SIZE := 65536
+        DEVICE_TITLE := EnGenius ENS620EXT
+        IMAGE_SIZE := 20439364
```
The firmware partition size is specified to be 0x14D0000 => 21823488

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -158,6 +158,22 @@ define Device/engenius_eap1300
+        DEVICE_DTS_CONFIG := config@4
+        BLOCKSIZE := 128k
+        PAGESIZE := 2048
+        DTB_SIZE := 65536
```
Is this DTB_SIZE needed by something, if not please remove it?

chunkeey commented on target/linux/ipq40xx/files-4.14/arch/arm/boot/dts/qcom-ipq4018-ens620ext.dts:
```diff
@@ -0,0 +1,277 @@
+		reg = <0>;
+		linux,modalias = "m25p80", "mx25l1606e", "n25q128a11";
+		spi-max-frequency = <24000000>;
+
```
If this is a SPI-NOR chip. Please consider testing if the ```m25p,fast-read;``` property is supported.

steveglennon replied:
Yes it looks like it does - will add to next commit.
Data sheet shows two max frequencies - 50MHz for read operations, 104MHz for everything else (including fast read). All other DTS files seem to choose 24MHz. Any insight as to whether 50MHz or 104MHz is the right choice for spi-max-frequency?

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -158,6 +158,22 @@ define Device/engenius_eap1300
+        DEVICE_DTS := qcom-ipq4018-ens620ext
+        DEVICE_DTS_CONFIG := config@4
+        BLOCKSIZE := 128k
+        PAGESIZE := 2048
```
This is weird, can you please provide what flash chip your device is using? Because if it's SPI-NOR then the BLOCKSIZE and PAGESIZE are probably wrong. If it's SPI-NAND then you need ubi.

steveglennon replied:
I have not looked directly on the PCB but the uboot log shows W25Q256 (SPI NOR) :
FLASH:  SF: Detected W25Q256 with page size 64 KiB, total 32 MiB
ipq_spi: page_size: 0x100, sector_size: 0x10000, size: 0x2000000
32 MB

Looks like BLOCKSIZE should be 64k and PAGESIZE should be 256.

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -158,6 +158,22 @@ define Device/engenius_eap1300
+        DTB_SIZE := 65536
+        DEVICE_TITLE := EnGenius ENS620EXT
+        IMAGE_SIZE := 20439364
+        KERNEL_SIZE := 5120k
```
Is this limit something the u-boot loader enforces? Can you please provide a "printenv" from the u-boot prompt / or a dump of your u-boot-env partition. Thanks.

steveglennon replied:
printenv: 
app_part=0
athaddr=88:dc:96:00:00:13
baudrate=115200
bootcmd=bootipq
bootdelay=2
bundle_ap_mac=88:dc:96:00:00:15
country=000
ddwrt_sw=0
debug=0
domain=0
enable_failsafe=0
ethact=eth0
ethaddr=88:dc:96:00:00:10
fwaddr=88:dc:96:00:00:14
hw_id=00000000
hw_ver=0.0.1
ipaddr=192.168.1.1
language_code=00
machid=8010000
oled_on=0
op_mode=0
pro_id=000
serverip=192.168.1.37
service_tag=000000000000000
sn=000000001
snextra=00000000000000000000
stderr=serial
stdin=serial
stdout=serial
uboot_ver=1.0.0.1
wanaddr=88:dc:96:00:00:11
wlanaddr=88:dc:96:00:00:12

chunkeey replied:
Thank you.

```bootcmd=bootipq```

so you can leave it as is.

chunkeey commented on target/linux/ipq40xx/image/Makefile:
```diff
@@ -158,6 +158,22 @@ define Device/engenius_eap1300
+        KERNEL_SIZE := 5120k
+        FILESYSTEMS := squashfs
+        IMAGES := sysupgrade.bin
+        IMAGE/sysupgrade.bin := append-kernel | pad-to $$$$(KERNEL_SIZE) | append-rootfs | pad-rootfs | append-metadata | check-size $$$$(IMAGE_SIZE)
```
The metadata should get removed by the sysupgrade. so please do the check-size before append-metadata.

(The pad-to looks wrong in this context together with the use of denx,fit splitter).

chunkeey reviewed (changes requested):
Here are a few notes that need to be addressed.

- I think your commit message has a stray "|" in the first line at the end? 

- please use sysupgrade "-n" switch in the initial installation instructions.
  The reason is that we want to have a clean state and not have the old
  vendors' configuration around, since this is bound to break one day.

- To make live easier for the Table Of Hardware maintainers, can you please a
  quick overview of the hardware into the commit messages on top? If you need
  an example, please look at the recent FritzBox 7530 patch:
  https://github.com/openwrt/openwrt/commit/95b0c07a618fe5fd93a26931152ced483bba143b

- (other patch related comments are inline)

Thanks.

steveglennon commented:
Addressed all the resolutions requested by @chunkeey. Let me know if I messed up.

steveglennon commented:
Hold please - the rebase didn't quite work.

chunkeey commented:
> Hold please - the rebase didn't quite work.

well, I pushed a version to my tree with some fixes.

https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=2970af4ff4302e574ea5437661f606aeb319d6bf

Is this good or is something missing? (i.e. kernel pad-to or firmware partition order, etc...)

steveglennon commented:
The 901 patch showed up as invalid - shame on me for not building to test following the rebase and before the push.

Just fixed that in my local copy, and also fixed the MAC addresses for wan (LAN1) and lan (LAN2) plus the wlan0 and wlan1 interfaces. Now extracts the base MAC address from u-boot-env.

I'm almost ready to push those.

steveglennon commented:
I should also add I am building and testing this with non-CT ath10k firmware. Have not tried the CT variant.

steveglennon commented:
OK, I think this is it. Thanks for your help @chunkeey and let me know if there is something else unacceptable.

chunkeey commented:
> OK, I think this is it. Thanks for your help @chunkeey and let me know if there is something else unacceptable.

You tell me:
https://git.openwrt.org/?p=openwrt/staging/chunkeey.git;a=commit;h=39b306f40260eaf862f9746bc5216bb37a3c1e57

steveglennon commented:
Built and installed. It all looks good.
Thanks

chunkeey commented:
Merged. Thanks.
